README
NAME

    Data::Transpose - iterate, filter and validate data, and transpose to
    different field names

DESCRIPTION

    Caters to your needs for manipulating data by different operations,
    which are filtering records, iterating records, validating and
    transposing to different field names.

VERSION

    Version 0.0022

SYNOPSIS

        use warnings;
        use strict;
        
        use Data::Transpose::Prefix;
        use Data::Dumper;
        
        my $data = {
            first => 'John',
            last  => 'Doe',
            foo   => 'bar',
        };
        
        my $dtp = Data::Transpose::Prefix->new(prefix => 'user.');
        foreach my $needs_prefix ( qw(first last) ) {
            $dtp->field( $needs_prefix );
        }
        
        my $output = $dtp->transpose( $data );
        
        print Data::Dumper->Dump([$data, $output], [qw(data output)]);

    outputs:

        $data = {
                  'first' => 'John',
                  'last' => 'Doe',
                  'foo' => 'bar'
                };
        $output = {
                    'user.last' => 'Doe',
                    'user.first' => 'John',
                    'foo' => 'bar'
                  };

REFERENCE

    Validator

      Data::Transpose::Validator

    Iterator

      Data::Transpose::Iterator

METHODS

 new

    Parameters for the constructor are:

    unknown

      Determines how to treat fields in the input hash which are not known
      to the Data::Transpose object:

      fail

	The transpose operation fails.

      pass

	Unknown fields in the input hash appear in the output hash. This is
	the default behaviour.

      skip

	Unknown fields in the input hash don't appear in the output hash.

      This doesn't apply to the "transpose_object" method.

 field

    Add a new field object and return it:

        $tp->field('email');

 group

    Add a new group object and return it:

        $tp->group('fullname', $tp->field('firstname'), $tp->field('lastname'));

 transpose

    Transposes input:

       $new_record = $tp->transpose($orig_record);

 transpose_object

    Transposes an object into a hash reference.

AUTHOR

    Stefan Hornburg (Racke), <racke at linuxia.de>

BUGS

    Please report any bugs or feature requests at
    https://github.com/racke/Data-Transpose/issues. I will be notified, and
    then you'll automatically be notified of progress on your bug as I make
    changes.

SUPPORT

    You can find documentation for this module with the perldoc command.

        perldoc Data::Transpose

    You can also look for information at:

      * Github's issue tracker (report bugs here)

      https://github.com/racke/Data-Transpose/issues

      * AnnoCPAN: Annotated CPAN documentation

      http://annocpan.org/dist/Data-Transpose

      * CPAN Ratings

      http://cpanratings.perl.org/d/Data-Transpose

      * Search CPAN

      http://search.cpan.org/dist/Data-Transpose/

ACKNOWLEDGEMENTS

    Todd Wade for mxcheck test failures fix (GH #5, #11) and improvements
    to documentation (GH #12, #13).

LICENSE AND COPYRIGHT

    Copyright 2012-2016 Stefan Hornburg (Racke).

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Data::Transpose

0.0022 Wed Feb 10 21:58:15 2016 CET

    [BUG FIXES]

    * Remove obsolete test credit card numbers for Discover cards
      (Slaven Rezić, Marco Pessotto, GH #25, #26).

0.0021 Tue Jan 19 11:49:42 2016 CET

    [BUG FIXES]

    * Do not rely on unknown domains for checking mail validation
      (Marco Pessotto, GH #24).

    [TESTS]

    * Add Travis CI configuration file (Stefan Hornburg/Racke).

0.0020 Sun Aug 16 10:53:30 2015 WEST

    [ENHANCEMENTS]

    * Add DataTime validator (Sam Batschelet, GH #16, #18).

    [BUG FIXES]

    * Avoid double errors on undefined strings in String validator
      (Marco Pessotto).

    [DOCUMENTATION]

    * Improve clarity in Data::Transpose::Validator::Subrefs
      (Peter Mottram, GH #19).
    * Remove superfluous "the" from POD for the Validator class
      (Stefan Hornburg).

    [TESTS]

    * Add tests and documentation for errors_hash method.

0.0015 Sun Jul  5 11:05:00 2015 CEST

    [BUG FIXES]

    * Fix "repetion" typo (Stefan Hornburg).

0.0014 Wed Jun  3 11:00:13 2015 CEST

    [BUG FIXES]

	* Fix stale data caused by missing input fields
	  (GH #14, Marco Pessotto).

    [DOCUMENTATION]

	* Add example to synopsis and improve name and
	  description (GH #12, #13, Todd Wade).

0.0013 Sun May 17 21:45:06 2015 CEST

    [TESTS]

    * Fix for mxcheck test failures (GH #5, #11, Todd Wade).

0.0012 Fri Feb  6 10:43:29 2015 CET

    [ENHANCEMENTS]

    * Add Data::Transpose::Prefix class (Stefan Hornburg).
    * Migrate Data::Transpose to Moo (GH #8, Marco Pessotto).
    * Add auto correction for email addresses with double dots
      before top level domain (Stefan Hornburg).

0.0011 Fri Dec  5 12:28:44 2014 CET

    [ENHANCEMENTS]

    * Added option collapse_whitespace to validator (GH #7, Marco Pessotto).
    * Clarify error message on hashref check in transpose method
      (Stefan Hornburg).

0.0010 Wed Apr 30 08:58:07 2014 CEST

    [BUG FIXES]

    * Remove the additional check on ' for email addresses, as this is a
      valid character per RFC 822 (GH #6, Marco Pessotto).

0.0009 Thu Mar 13 19:34:00 2014 CET

    * Add iterator Data::Transpose::Iterator::Scalar.

0.0008 Fri Feb 28 16:32:04 2014 CET

    * Add transpose_object method to Data::Transpose.

    * Add links to classes in POD for Data::Transpose.

0.0007 Tue Feb 25 19:16:45 2014 CET

    * Allow to pass validator options inside field arguments for named
      validators.

    * Fix typo in Data::Transpose::Validator's documentation for prepare
      method.

    * Fix typos in NumericRange iterator.

    * Fix link to credit card validator in the documentation.

    * Add Moo to prequisites.

0.0006 Fri Feb  7 17:33:56 2014 CET

    * Add base and errors iterators.

0.0005 Sat Jan 18 13:41:32 2014 CET

    * Add group validation.

    * Add success method to Data::Transpose::Validator.

    * Fix spurious =item in Data::Transpose::Validator's POD.

    * Run POD tests only when RELEASE_TESTING is set.

    * Add resource and bugtracker to Makefile.PL.

0.0004 Thu Feb 21 13:46:56 2013 CET

    * Added Data::Transpose::Validator, a module to validate incoming
      data (form-oriented).

0.0003

    * Added documentation

0.0002

    * Added Data::Transpose::EmailValid and Data::Transpose::PasswordPolicy

0.0001

    * Initial release.




lib/Data/Transpose.pm
package Data::Transpose;

use 5.010001;
use strict;
use warnings;

use Data::Transpose::Field;
use Data::Transpose::Group;

use Moo;
use MooX::Types::MooseLike::Base qw(:all);
use namespace::clean;

=head1 NAME

Data::Transpose - iterate, filter and validate data, and transpose to
different field names

=head1 DESCRIPTION

Caters to your needs for manipulating data by different operations,
which are filtering records, iterating records, validating and
transposing to different field names.

=head1 VERSION

Version 0.0022

=cut

our $VERSION = '0.0022';

=head1 SYNOPSIS

    use warnings;
    use strict;
    
    use Data::Transpose::Prefix;
    use Data::Dumper;
    
    my $data = {
        first => 'John',
        last  => 'Doe',
        foo   => 'bar',
    };
    
    my $dtp = Data::Transpose::Prefix->new(prefix => 'user.');
    foreach my $needs_prefix ( qw(first last) ) {
        $dtp->field( $needs_prefix );
    }
    
    my $output = $dtp->transpose( $data );
    
    print Data::Dumper->Dump([$data, $output], [qw(data output)]);

outputs:

    $data = {
              'first' => 'John',
              'last' => 'Doe',
              'foo' => 'bar'
            };
    $output = {
                'user.last' => 'Doe',
                'user.first' => 'John',
                'foo' => 'bar'
              };

=head1 REFERENCE

=over 4

=item Validator

L<Data::Transpose::Validator>

=item Iterator

L<Data::Transpose::Iterator>

=back

=head1 METHODS

=head2 new

Parameters for the constructor are:

=over 4

=item unknown

Determines how to treat fields in the input hash
which are not known to the Data::Transpose object:

=over 4

=item fail

The transpose operation fails.

=item pass

Unknown fields in the input hash appear in the output
hash. This is the default behaviour.

=item skip

Unknown fields in the input hash don't appear in
the output hash.

=back

This doesn't apply to the L</transpose_object> method.

=back

=cut

has unknown => (is => 'ro',
                isa => sub {
                    my $unknown = $_[0];
                    my %permitted = (
                                     fail => 1,
                                     pass => 1,
                                     skip => 1,
                                    );
                    die "unknown accepts only " . join(' ', keys %permitted)
                      unless $permitted{$unknown};
                },
                default => sub { 'pass' });

has _fields => (is => 'ro',
               isa => ArrayRef[Object],
               default => sub { [] },
              );


=head2 field

Add a new L<field|Data::Transpose::Field> object and return it:

    $tp->field('email');

=cut

sub field {
    my ($self, $name) = @_;
    my ($object);

    $object = Data::Transpose::Field->new(name => $name);

    push @{$self->_fields}, $object;

    return $object;
}

=head2 group

Add a new L<group|Data::Transpose::Group> object and return it:

    $tp->group('fullname', $tp->field('firstname'), $tp->field('lastname'));

=cut

sub group {
    my ($self, $name, @objects) = @_;
    
    my $object = Data::Transpose::Group->new(name => $name,
                                             objects => \@objects);

    push @{$self->_fields}, $object;
    
    return $object;
}

=head2 transpose

Transposes input:

   $new_record = $tp->transpose($orig_record);

=cut

sub transpose {
    my ($self, $vref) = @_;
    my ($weed_value, $fld_name, $new_name, %new_record, %status);

    $status{$_} = 1 for keys %$vref;

    for my $fld (@{$self->_fields}) {
        $fld_name = $fld->name;

        # set value and apply operations
        if (exists $vref->{$fld_name}) {
            $weed_value = $fld->value($vref->{$fld_name});
        }
        else {
            $weed_value = $fld->value(undef);
        }

        if ($new_name = $fld->target) {
            $new_record{$new_name} = $weed_value;
        }
        else {
            $new_record{$fld_name} = $weed_value;
        }

        delete $status{$fld_name};
    }

    if (keys %status) {
        # unknown fields
        if ($self->unknown eq 'pass') {
            # pass through unknown fields
            for (keys %status) {
                $new_record{$_} = $vref->{$_};
            }
        }
        elsif ($self->unknown eq 'fail') {
            die "Unknown fields in input: ", join(',', keys %status), '.';
        }
    }

    return \%new_record;
}

=head2 transpose_object

Transposes an object into a hash reference.

=cut

sub transpose_object {
    my ($self, $obj) = @_;
    my ($weed_value, $fld_name, $new_name, %new_record, %status);

    for my $fld (@{$self->_fields}) {
        $fld_name = $fld->name;

        # set value and apply operations
        if ($obj->can($fld_name)) {
            $weed_value = $fld->value($obj->$fld_name());
        }
        else {
            $weed_value = $fld->value;
        }

        if ($new_name = $fld->target) {
            $new_record{$new_name} = $weed_value;
        }
        else {
            $new_record{$fld_name} = $weed_value;
        }
    }

    return \%new_record;
}

=head1 AUTHOR

Stefan Hornburg (Racke), C<< <racke at linuxia.de> >>

=head1 BUGS

Please report any bugs or feature requests at
L<https://github.com/racke/Data-Transpose/issues>.
I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Data::Transpose

You can also look for information at:

=over 4

=item * Github's issue tracker (report bugs here)

L<https://github.com/racke/Data-Transpose/issues>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Data-Transpose>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Data-Transpose>

=item * Search CPAN

L<http://search.cpan.org/dist/Data-Transpose/>

=back


=head1 ACKNOWLEDGEMENTS

Todd Wade for mxcheck test failures fix (GH #5, #11)
and improvements to documentation (GH #12, #13).

=head1 LICENSE AND COPYRIGHT

Copyright 2012-2016 Stefan Hornburg (Racke).

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

1; # End of Data::Transpose
